bazel-mode.el

```diff
@@ -0,0 +1,167 @@
+
+(defconst common-testrule-args
+  '("local" "flaky" "shard_count" "size" "timeout"))
+
+(defconst config-setting-args
+  '("values"))
+
+(defconst genrule-args
+  '("cmd" "compatible_with" "deprecation" "distribs" "executable" "features"
+	"licenses" "message" "name" "output_licenses" "output_to_bindir" "outs"
+	"restricted_to" "srcs" "tags" "testonly" "tools" "visibility"))
+
+(defconst genquery-args
+  '("deps" "data" "expression" "opts" "scope" "strict"))
+
+(defconst git-repository-args
+  '("commit" "init_submodules" "remote" "sha256"))
+
+(defconst glob-args
+  '("exclude" "exclude_directories" "include"))
+
+(defconst http-args
+  '("executable" "strip_prefix" "type" "url" "urls"))
+
+(defconst java-buildrule-args
+  '("classpath_resources" "create_executable" "deploy_manifest_lines"
+	"javacopts" "jvm_flags" "launcher" "main_class" "output_licenses" "plugins"
+	"resources" "resource_jars" "resource_strip_prefix" "restricted_to"
+	"runtime_deps" "use_testrunner"))
+
+(defconst local-repository-args
+  '("path"))
+
+(defconst maven-args
+  '("artifact" "repository" "server" "sha1"))
+
+(defconst new-repository-args
+  '("build_file" "build_file_content" "workspace_file" "workspace_file_content"))
+
+(defconst package-args
+  '("default_deprecation" "default_testonly" "default_visibility"))
+
+(defconst py-buildrule-args
+  '("default_python_version" "default_runtime" "files" "imports" "interpreter"
+    "interpreter_path" "main" "protoc" "srcs_version"))
+
+(defconst bazel-types
+  (bazel-mode--sort-strings-longest-first
+    (append bind-args cc-buildrule-args common-buildrule-args common-testrule-args
+	  config-setting-args genrule-args genquery-args git-repository-args glob-args
+	  http-args java-buildrule-args local-repository-args maven-args
+	  new-repository-args package-args py-buildrule-args)))
+
+(defconst bazel-constants
+  '("true"))
+
+(defconst bazel-events
+  '())
+
+;; generate regex string for each category of keywords
+(setq bazel-keywords-regexp (regexp-opt bazel-keywords 'words))
+(setq bazel-functions-regexp (regexp-opt bazel-functions 'words))
+(setq bazel-type-regexp (regexp-opt bazel-types 'words))
+(setq bazel-constant-regexp (regexp-opt bazel-constants 'words))
+(setq bazel-event-regexp (regexp-opt bazel-events 'words))
+
+(defvar bazel-font-lock-keywords
+      `(
+        (,bazel-keywords-regexp . font-lock-keyword-face)
+        (,bazel-functions-regexp . font-lock-function-name-face)
+        (,bazel-type-regexp . font-lock-type-face)
+        (,bazel-constant-regexp . font-lock-constant-face)
+        (,bazel-event-regexp . font-lock-builtin-face)
+        ;; note: order above matters. in general, longer words first.
+        ))
+
+(defvar bazel-mode-syntax-table
+  (let ((table (make-syntax-table)))
+	;; comments start with '#' and end with line break
+	(modify-syntax-entry ?# "<" table)
+	(modify-syntax-entry ?\n ">" table)
+	table))
+
+;;;###autoload
+(define-derived-mode bazel-mode prog-mode "Bazel"
```

Comment:
I think it's better for readers to reorder functions from high-level to low-level. That way readers see the interface of the package first.

Reply:
Done.